File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/BinaryTree/pathSum_113.java
package algo.tzashinorpu.SpecialSubject.BinaryTree;

import algo.tzashinorpu.SpecialSubject.TreeNode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class pathSum_113 {
    List<List<Integer>> res;

    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        List<Integer> list = new ArrayList<>();
        res = new ArrayList<>();
        helper(root, list, targetSum);
        return res;
    }

    private void helper(TreeNode root, List<Integer> list, int targetSum) {
        if (root == null) return;
        if (root.left == null && root.right == null) {
            if (targetSum == root.val) {
                list.add(root.val);
                ArrayList<Integer> clone = new ArrayList<>(list);
                res.add(clone);
                list.remove(list.size() - 1);
            }
            return;
        }
        list.add(root.val);
        helper(root.left, list, targetSum - root.val);
        helper(root.right, list, targetSum - root.val);
        list.remove(list.size() - 1);
    }
}

File: Clion/ZeroTrac/Tree/pathSum_113.cpp
//
// Created by TzashiNorpu on 10/25/2022.
//
#include "tree.h"

using namespace Tree;

vector<vector<int>> Solution::pathSum(TreeNode *root, int targetSum) {
  vector<vector<int>> res;
  vector<int> tmp;
  int sum = 0;
  function<void(TreeNode *)> dfs = [&](TreeNode *node) -> void {
      if (!node) return;
      if (!node->left && !node->right)
        if (sum + node->val == targetSum) tmp.push_back(node->val), res.push_back(tmp), tmp.pop_back();
      sum += node->val;
      tmp.push_back(node->val);
      dfs(node->left);
      dfs(node->right);
      sum -= node->val;
      tmp.pop_back();
  };
  dfs(root);
  return res;
}